### ISSUE TITLE
ToolsPanel: Selected ToolsPanel menu options are not persistent

### ISSUE THREAD

**Issue by apeatling** [WordPress/gutenberg#41544]
Prior: https://github.com/WordPress/gutenberg/issues/41376

When using the ToolsPanel menu to select design tool options for a block, these selections are not remembered unless the value of a design tool setting has been selected. Simply setting the option as visible is not enough to persist it on block re-selection.

This GIF shows the issue:

![2022-06-03 10 08 20](https://user-images.githubusercontent.com/1464705/171912749-75525de3-ea51-4a50-8cb9-8311e05f2a7b.gif)

**Comment by apeatling** [comment]
Persisting selected options has been discussed previously in https://github.com/WordPress/gutenberg/pull/39561#discussion_r830803440. 

However this discussion was for persisting selected options across multiple blocks of the same type. That warrants caution, however persisting options in the same instance of a block is quite different. In my opinion this goes against user intent and expectation and should be added.

**Comment by mtias** [comment]
Curious how you'd expect it working after a reload — would you expect a non customized option you selected to show up? Holding onto the session state seems a bit superfluos to me, and I don't think it's worth the extra complexity of handling state that would persist through remounts but not through hard reloads.

**Comment by apeatling** [comment]
I don't think it is necessary to maintain state after a reload. 

However, within a single session, I can see it being a confusion. I can see a user selecting the tools they need, but in the middle selecting a similar block to refer to a setting value. Then they come back to the block to find everything reset.

I haven't run in to this in my real world usage, so it is purely hypothetical on my part -- but it has been reported in the original issue.

**Comment by mtias** [comment]
I'd consider this low impact for the likely complexity it'd entail to keep track of unmounted component state. We can revisit if it becomes a wider concern.

**Comment by annezazu** [comment]
Re-opening this as it [came up in a discussion with YouTubers](https://www.youtube.com/live/qLOUBibVhKw?si=elpwnFOq4bAgUGb4&t=2879) with folks requesting this as an option. Specifically, that when an option is enabled that it persists in the future until someone disables it again for typography settings. I recorded a little video showing how this impacts my workflow with a specific example:

https://github.com/user-attachments/assets/3e31481a-344d-4665-b9b9-21b039391341

Curious to hear from @WordPress/outreach as to whether this is something others run into.

**Comment by markhowellsmead** [comment]
Definitely. See also #62064.

**Comment by bacoords** [comment]
Big agree on this. There's a number of issues about the persistence of UI controls- it's a very common pain point for most users (There were some a11y concerns about some of them, but I believe they're addressable): 

- https://github.com/WordPress/gutenberg/issues/51104
- https://github.com/WordPress/gutenberg/issues/62064
- https://github.com/WordPress/gutenberg/issues/45283
- https://github.com/WordPress/gutenberg/issues/59789

**Comment by annezazu** [comment]
Thank you all for chiming in and cross connecting. Excellent to get a broader spread and more thoughts. Keep it coming!

**Comment by markhowellsmead** [comment]
There's some great UX feedback in this post on X. What's the best way to get these included here in Github? 

https://x.com/KatieKeithBarn2/status/1824362828251201966

**Comment by annezazu** [comment]
>There's some great UX feedback in this post on X. What's the best way to get these included here in Github?

Open or find issues for each item :) Some of these already have issues opened.

**Comment by annezazu** [comment]
Noting that this came up once more today on a WordPress freelancer online meetup with folks talking about how frustrating it is to go looking for an option, turn it on, and then have it disappear when you move away to work on another block causing you to have to repeat the search cc @techmagick as you were a part of that convo <3

**Comment by ciampo** [comment]
Hey folks, in the context of [potentially stabilizing the `ToolsPanel` component](https://github.com/WordPress/gutenberg/issues/67954), I was reviewing related opened issues.

For this issue specifically, I think we need to make changes to both the `ToolsPanel` component itself, and to the logic in the inspector control sidebar:

- we may need to add functionality to the component to:
  - add an prop to set an _optional_  `ToolsPanelItem` as visible (I believe that currently only "default" items are visible on first render)
  - add a callback prop so that consumers of `ToolsPanel` can be notified of which panels are shows and which are hidden
- in the inspector controls (or wherever the component is used), the information about which `ToolsPanelItem` was shows for each block can be stored and used to make such items persistent.

@WordPress/gutenberg-components  @aaronrobertshaw how does this sound?

**Comment by aaronrobertshaw** [comment]
> how does this sound?

Thanks for kicking off this discussion again @ciampo!

As per this [comment](https://github.com/WordPress/gutenberg/issues/41544#issuecomment-1242733713), I'm not sure we've reached the threshold where the added complexity in tracking menu states is desirable. However, if it doesn't take too much to add the new functionality to the `ToolsPanel` component to support this in the future, then the plan sounds good to me and unblocks the proposed stabilization of the component 👍

**Comment by fabiankaegy** [comment]
@aaronrobertshaw @ciampo in chatting with @youknowriad a bit over the last week we thought about it this way:

 1. Refactor any existing instance of the `PanelBody` in Gutenberg to use the `ToolsPanel` (https://github.com/WordPress/gutenberg/issues/67813)
 2. Standardize the usage of `group` on all tool panels to make it easier to expend these panels for third parties (https://github.com/WordPress/gutenberg/issues/67814)
 3. Ensure that any tool panel items tie into the `defaultControls` mechanism so we can control the visibility of all items (https://github.com/WordPress/gutenberg/pull/67729)
 4. potentially connect that default visibility to user settings so that their choices can be persisted

Does that make sense or are we missing something in there?

**Comment by aaronrobertshaw** [comment]
> Standardize the usage of group on all tool panels to make it easier to expend these panels for third parties (https://github.com/WordPress/gutenberg/issues/67814)

My initial impression here is this is back to front. Maybe my brain is fried after a long day so correct me if I'm wrong.

Group isn't really "used on tools panels". We have slot fills representing these groups. Those slots can render however they want and so group slots representing the block supports render their fills wrapped in a `ToolsPanel`. It sounds like #67814 wants to create a range of new slots for both common (settings) and random one-off use cases (filters).

This all sounds separate to the ToolsPanel component to me.

> Ensure that any tool panel items tie into the defaultControls mechanism so we can control the visibility of all items (https://github.com/WordPress/gutenberg/pull/67729)

The broader mechanism for `defaultControls` shouldn't impact the `ToolsPanel` component. It's up to the consumer to say whether a `ToolsPanelItem` should be displayed by default or not.

> potentially connect that default visibility to user settings so that their choices can be persisted

@ciampo's [proposal](https://github.com/WordPress/gutenberg/issues/41544#issuecomment-2545016098) above would make this possible. With the increased proliferation of the `ToolsPanel` due to efforts such as #67813, it means we could have heaps of panels to have to track within user settings. This is where the [concern regarding impact vs complexity](https://github.com/WordPress/gutenberg/issues/41544#issuecomment-1242733713) comes in.

**Comment by fabiankaegy** [comment]
@aaronrobertshaw yeah sorry my numbered list really made it look like those steps need to happen in that order 🤦 

I don't think any of these really depend on one another. 

All I was trying to say is that maybe we should leverage the `defaultControls` work you are doing to persist which items are shown and which items are hidden here in the tools panels

**Comment by aaronrobertshaw** [comment]
> All I was trying to say is that maybe we should leverage the defaultControls work you are doing to persist which items are shown and which items are hidden here in the tools panels

There might be some limited overlap between `defaultControls` and the persisting of menu options. I'm not 100% convinced yet whether that would extend beyond similar shapes of data.

Ultimately, the `defaultControls` mechanism isn't something that will persist these user-made selections. There'll have to be a separate feature to collect, persist, and merge them with `defaultControls`. In other words, there's not much that could be leveraged towards the new feature other than the `defaultControls` data itself as a base.

**Comment by fabiankaegy** [comment]
@aaronrobertshaw thanks for those insights :) That makes perfect sense. 

I think both here and in https://github.com/WordPress/gutenberg/issues/67813 I have been too implementation focussed. 

At the end of the day, these are my feature goals that I am completely open on how we get there:

 - All Block Settings & Styles should be grouped in relevant groups 
 - These groups would be standardizes as much as possible. 
 - These groups should all use the same UI
 - Developers should be able to place additional components in any of these groups
 - There should be a mechanism to control which controls are shown by default in any of the groups
 - Users should be able to have the controls they want to see by default override what the developer thinks they should see by default

**Comment by ciampo** [comment]
Thank you all!

If I understand everything correctly, a lot of the features being discussed are not directly affecting `ToolsPanel`, and therefore the changes suggested [above](https://github.com/WordPress/gutenberg/issues/41544#issuecomment-2545016098) should still be enough, from the component's POV ?

**Comment by aaronrobertshaw** [comment]
That's my understanding @ciampo, your proposed changes cover what's needed from the component itself.

**Comment by ciampo** [comment]
Cool! Let's use https://github.com/WordPress/gutenberg/issues/67954 to coordinate next steps and related timelines.
